### PR TITLE
read manifest from SmartWeave contract

### PR DESCRIPTION
**Pull request by ppedziwiatr**
related to https://github.com/redstone-finance/redstone-smartweave-contracts/pull/1

### CODE REVIEW

File: src/NodeRunner.ts [hunk 1]
```diff
@@ -11,35 +11,37 @@ import mode from "../mode";
...
 import PricesService, {PricesBeforeAggregation, PricesDataFetched} from "./fetchers/PricesService";
-import {mergeObjects} from "./utils/objects";
-import _ from "lodash";
+import {mergeObjects, readJSON, sleep} from "./utils/objects";
+import ManifestConfigError from "./manifest/ManifestConfigError";
 
 const logger = require("./utils/logger")("runner") as Consola;
 const pjson = require("../package.json") as any;
 
+export const MANIFEST_REFRESH_INTERVAL = 30000;
```
Comment (ppedziwiatr) [inline src/NodeRunner.ts:20]: not sure how often we should be checking for new manifest..

File: src/NodeRunner.ts [hunk 2]
```diff
@@ -126,45 +153,45 @@ export default class NodeRunner {
...
-    await this.broadcastPrices(signedPrices)
-
-    await this.broadcastEvmPricePackage(signedPrices);
+    if (!this.manifest!.developmentMode) {
+      await NodeRunner.broadcastPrices(signedPrices);
+      await this.broadcastEvmPricePackage(signedPrices);
 
-    if (mode.isProd) {
-      await this.arService.storePricesOnArweave(arTransaction);
+      if (mode.isProd) {
```
Comment (jakub-wojciechowski) [inline src/NodeRunner.ts:164]: Now we've got two models (isProd and developmentMode), let's discuss tomorrow how to consistently handle dev mode (maybe using two flags disableArweaveStorage & disableBroadcasting)

File: src/NodeRunner.ts [hunk 3]
```diff
@@ -217,21 +252,47 @@ export default class NodeRunner {
...
+    const now = Date.now();
+    const timeDiff = now - this.lastManifestLoadTimestamp!;
+
+    logger.info("Time since last check:", {timeDiff, "manifestRefreshInterval": MANIFEST_REFRESH_INTERVAL})
+
+    if (timeDiff >= MANIFEST_REFRESH_INTERVAL) {
+      logger.info("Trying to fetch new manifest version.");
+      const manifestFetchTrackingId = trackStart("Fetching manifest");
+      try {
+        const manifest = await this.arweaveService.getCurrentManifest();
```
Comment (jakub-wojciechowski) [inline src/NodeRunner.ts:274]: How long does it take to load manifest?
The current approach with waiting could be problematic for short-interval manifests if we use 30s interval. 
What do you think about non-blocking way of loading manifest (we load in a non-blocking promise and set a flag if it's different. Then in every step we first check the flag and then try to load if manifest refresh interval is reached. 
We can discuss it and implement together with manifestService refactor. 
(BTW, we should also check how the ar balance checking could affect our price fetching for short intervals).

File: src/NodeRunner.ts [hunk 4]
```diff
@@ -48,18 +50,37 @@ export default class NodeRunner {
...
+        try {
+          manifestData = await arweaveService.getCurrentManifest();
+        } catch (e) {
+          logger.error("Initial manifest read failed.", e.stack || e);
+        }
+        if (manifestData !== null) {
+          logger.info("Fetched manifest", manifestData)
+          break;
+        }
+        await sleep(2000);
```
Comment (jakub-wojciechowski) [inline src/NodeRunner.ts:73]: It's not a big deal, but I'm wondering what's the added value of waiting here (I don't think there is any rate-limiting mechanism on smartweave).

**Review by jakub-wojciechowski — APPROVED** [review]

File: src/NodeRunner.ts (hunk 4 above)
Comment (ppedziwiatr) [reply src/NodeRunner.ts:73]: removing.

File: src/NodeRunner.ts (hunk 3 above)
Comment (ppedziwiatr) [reply src/NodeRunner.ts:274]: well, I've understood that the whole process was supposed to be as simple as possible, but of course I can introduce a mechanism with a separate Promise and additional flag.

File: src/NodeRunner.ts (hunk 2 above)
Comment (ppedziwiatr) [reply src/NodeRunner.ts:164]: I'm removing developmentMode for now.

File: src/NodeRunner.ts (hunk 3 above)
Comment (ppedziwiatr) [reply src/NodeRunner.ts:274]:
```
2021-06-16T19:05:21.855+02:00 | {"date":"2021-06-16T17:05:21.853Z","args":["Trying to fetch new manifest version."]

2021-06-16T19:05:27.612+02:00 | {"uploadBlockHeight":708058,"manifestTxId":"LhBK42S69l3ho4ABSO-IvwAcbvYObt99vcxbe3Ugiaw"
```

it usually takes about 5-6 seconds.






2021-06-16T19:05:21.855+02:00 | {"date":"2021-06-16T17:05:21.853Z","args"
-- | --




2021-06-16T19:05:27.612+02:00 | {"uploadBlockHeight":708058,"manifestTxId":"LhBK42S69l3ho4ABSO-IvwAcbvYObt99vcxbe3Ugiaw",
-- | --
```

**Pull request merged**